assets/src/features/room-page/utils/usePinning.tsx

```diff
@@ -0,0 +1,46 @@
+import { useState } from "react";
+
+export type PinningApi = {
+  pinnedTileIds: string[];
+  pin: (tileId: string) => void;
+  unpin: (tileId: string) => void;
+  pinIfNotAlreadyPinned: (tileId: string) => void;
+};
+
+const usePinning = (): PinningApi => {
+  const [pinnedTileIds, setPinnedTileIds] = useState<string[]>([]);
+  const [pinnedTileIdHistory, setPinnedTileIdHistory] = useState<string[]>([]);
+  const historyLimit = 20;
+
+  const pin = (newTileId: string) => {
+    const updateHistory = (tileId: string) => {
+      setPinnedTileIdHistory((oldHistory) => {
+        if (oldHistory.length > historyLimit) oldHistory.pop();
+        return oldHistory.includes(tileId) ? oldHistory : [tileId, ...oldHistory];
+      });
+    };
```

Comment:
You don't need to limit the history size. You could use `Set` instead of an array. 

Also `.pop()` function mutates an array and it's not permitted in `React.useState()`